### ISSUE TITLE
Resistances from PowerPlay persist across class change

### ISSUE THREAD

**Issue by NoctisUGC** [Phil25/RTD#79]
We have been experiencing some issues with players getting insane resistance when they are not in an RTD state. I believe it could be a god-mode bug. This has never happened before until the new RTD update was added. Any chance for this to be looked into?

**Comment by Phil25** [comment]
Do you happen to have server logs or experience any errors? This could be related to #73.

**Comment by NoctisUGC** [comment]
Unfortuently we do not have any logs or errors. But my moderators have provided me with video recordings on the all the prev issues with what we think is RTD 

The latest issue regarding it - https://www.mediafire.com/file/3p0y0nqqfbzy6bb/wesssssss.mp4/file

Another issue regarding little to no damage (Date was 1st December) - https://www.mediafire.com/file/11p3618jfzxt2xv/wtfnodamage.mp4/file

"Wes" rolled sickness but somehow had insane health regen. Not sure if its an exploit or an RTD bug - https://www.mediafire.com/file/3p0y0nqqfbzy6bb/wesssssss.mp4/file

Image provided shows what a user did before the issue occurred. - 
![image](https://github.com/Phil25/RTD/assets/125376910/272857ea-a06c-43b8-a016-3876a41eb0bb)


Will upload more if we come across it again or more videos

**Comment by Phil25** [comment]
Thank you that’s a lot of help. There was one exploit fixed in 2.4.0 which involved you changing classes outside of spawn to get out of bad perks. There may have been a bug introduced because of it.

I will investigate this but unfortunately I’m currently away on vacation so I’d have to get back to it in a week.

**Comment by NoctisUGC** [comment]
Yeah, no worries. Thank you

**Comment by NoctisUGC** [comment]
> Thank you that’s a lot of help. There was one exploit fixed in 2.4.0 which involved you changing classes outside of spawn to get out of bad perks. There may have been a bug introduced because of it.
> 
> I will investigate this but unfortunately I’m currently away on vacation so I’d have to get back to it in a week.

I was looking into reverting the powerplay/godmode back to its original code for the servers until a fix could be added.

I came across this:

"switch (TF2_GetPlayerClass(client))
	{
		case TFClass_Scout:
		{
			SDKUnhook(client, SDKHook_OnTakeDamage, PowerPlay_ResistanceAndSlowdown);
			TF2Attrib_RemoveByDefIndex(client, Attribs.JumpHeight);
			SetSpeed(client, Cache[client].BaseSpeed);
		}

		case TFClass_Heavy:
		{
			SDKUnhook(client, SDKHook_OnTakeDamage, PowerPlay_Resistance);
		}

		default:
		{
			SDKUnhook(client, SDKHook_OnTakeDamage, PowerPlay_Resistance);
			TF2Attrib_RemoveByDefIndex(client, Attribs.ForceDamageTaken);
		}
	}"
	
	What would happen if you put break statements there? so there isn't a fall-through?

**Comment by Phil25** [comment]
SourcePawn switches don’t fallthrough, only one case is ever executed.

I’m guessing the issue comes from how prevention against switching classes to get out of bad perks was implemented (https://github.com/Phil25/RTD/commit/4724196033db0913f58aa9ea13fa1eb7e3d4a96a).

**Comment by Phil25** [comment]
Upon investigation I've confirmed this comes from PowerPlay and how class switch detection has changed. Namely, PowerPlay is trying to remove resistances for the class that the player is currently in, which now might differ from the class that the perk was applied for. This has been fixed by caching what class the perk has been applied with and working on that alone.

This also had a chance to cause issues for Infinite Ammo albeit without any consequences.

I've fixed it locally and will prepare a release shortly, but first I want to investigate other issues that have been reported.
